quil-rs/src/program/mod.rs

```diff
@@ -348,6 +356,110 @@ impl Program {
         Ok(expanded_program)
     }
 
+    /// Resolve [`LabelPlaceholder`]s and [`QubitPlaceholder`]s within the program using default resolvers.
+    ///
+    /// See [`resolve_placeholders_with_custom_resolvers`](Self::resolve_placeholders_with_custom_resolvers),
+    /// [`default_target_resolver`](Self::default_target_resolver),
+    /// and [`default_qubit_resolver`](Self::default_qubit_resolver) for more information.
+    pub fn resolve_placeholders(&mut self) {
+        self.resolve_placeholders_with_custom_resolvers(
+            self.default_target_resolver(),
+            self.default_qubit_resolver(),
+        )
+    }
+
+    /// Resolve [`TargetPlaceholder`]s and [`QubitPlaceholder`]s within the program such that the resolved values
+    /// will remain unique to that placeholder within the scope of the program.
+    ///
+    /// The provided `target_resolver` and `qubit_resolver`, will be used to resolve those values respectively.
+    /// If your placeholder returns `None` for a particular placeholder, it will not be replaced but will be left as a placeholder.
+    ///
+    /// If you wish to provide a resolver for either labels or qubits, but want to rely on the
+    /// default behavior for the other, considering using either
+    /// [`default_qubit_resolver`](Self::default_qubit_resolver) or [`default_target_resolver`](Self::default_target_resolver).
+    #[allow(clippy::type_complexity)]
+    pub fn resolve_placeholders_with_custom_resolvers(
+        &mut self,
+        target_resolver: Box<dyn Fn(&TargetPlaceholder) -> Option<String>>,
+        qubit_resolver: Box<dyn Fn(&QubitPlaceholder) -> Option<u64>>,
```

Comment:
Is there a specific reason for using `Box<dyn Fn()>` instead of `impl Fn()`?

Reply:
Not sure, @kalzoo do you remember why you went this route?